pkgs/build-support/setup-hooks/audit-tmpdir.sh

```diff
@@ -15,25 +15,50 @@ auditTmpdir() {
 
     echo "checking for references to $TMPDIR/ in $dir..."
 
-    local i
-    find "$dir" -type f -print0 | while IFS= read -r -d $'\0' i; do
-        if [[ "$i" =~ .build-id ]]; then continue; fi
+    local tmpdir="$(mktemp -d)"
+    local elf_fifo="$tmpdir/elf"
+    local script_fifo="$tmpdir/script"
+    mkfifo "$elf_fifo" "$script_fifo"
 
-        if isELF "$i"; then
-            if { printf :; patchelf --print-rpath "$i"; } | grep -q -F ":$TMPDIR/"; then
-                echo "RPATH of binary $i contains a forbidden reference to $TMPDIR/"
+    # Classifier: identify ELF and script files
+    (
+        find "$dir" -type f -not -path '*/.build-id/*' -print0 \
+        | while IFS= read -r -d $'\0' file; do
+            if isELF "$file"; then
+                printf '%s\0' "$file" >&3
+            elif isScript "$file"; then
+                if [ -e "$(dirname "$file")/.$(basename "$file")-wrapped" ]; then
+                    printf '%s\0' "$file" >&4
+                fi
+            fi
+        done
+        exec 3>&- 4>&-
+    ) 3> "$elf_fifo" 4> "$script_fifo" &
+
+    # Handler: check RPATHs concurrently
+    (
+        xargs -0 -P "$NIX_BUILD_CORES" -n 1 sh -c '
+            if { printf :; patchelf --print-rpath "$1"; } | grep -q -F ":$TMPDIR/"; then
+                echo "RPATH of binary $1 contains a forbidden reference to $TMPDIR/"
                 exit 1
             fi
-        fi
+        ' _ < "$elf_fifo"
+    ) &
+    local pid_elf=$!
 
-        if isScript "$i"; then
-            if [ -e "$(dirname "$i")/.$(basename "$i")-wrapped" ]; then
-                if grep -q -F "$TMPDIR/" "$i"; then
-                    echo "wrapper script $i contains a forbidden reference to $TMPDIR/"
-                    exit 1
-                fi
+    # Handler: check wrapper scripts concurrently
+    local pid_script
+    (
+        xargs -0 -P "$NIX_BUILD_CORES" -n 1 sh -c '
+            if grep -q -F "$TMPDIR/" "$1"; then
+                echo "wrapper script $1 contains a forbidden reference to $TMPDIR/"
+                exit 1
             fi
-        fi
+        ' _ < "$script_fifo"
+    ) &
+    local pid_script=$!
+
+    wait "$pid_elf" "$pid_script"
```

Comment:
Because we now print other output potentially after the error, we should do something like this:

```suggestion
    wait "$pid_elf" || { echo "Some binaries contain forbidden references to $TMPDIR/. Check the error above!"; exit 1 }}  
    wait "$pid_script" || { echo "Some scripts contain forbidden refences to $TMPDIR/. Check the error above!"; exit 1 }}  
```

Reply:
What's the benefit of this? The only case where the processes would return negative is when they already printed an error. I see no point in re-printing.

Reply:
The idea was if it prints multiple things that might look like an error it might not be clear which one caused the build to fail. Usually a user expects the last error to be the root cause but here we delay the termination so this might be no longer the case. This way we make sure to not send people down the wrong rabbit hole.